File: py/src/axon/hooks/hosts/codex.py
"""The Codex CLI codec. Codex's handler contract mirrors Claude Code's:
stdin JSON envelope, stdout JSON decision, exit code. Codex's host-side
event names also equal the canonical names.

Mirrors hooks/hosts/codex/codec.go; matched against
ts/src/hooks/hosts/codex.ts for shape.
"""

from __future__ import annotations

from axon.capabilities import Capabilities, level, load_capabilities
from axon.errors import SchemaError, UnsupportedActionError, UnsupportedEventError
from axon.hooks.decision import Decision
from axon.hooks.input import Input
from axon.registry import get

_KNOWN_INPUT_KEYS = {"hook_event_name", "session_id", "cwd", "tool_name", "tool_input", "tool_response", "prompt"}

_NATIVE_ACTION = {"allow": "allow", "warn": "ask", "block": "deny"}


class CodexCodec:
    def __init__(self) -> None:
        self._caps = load_capabilities("codex")
        self._host_to_canonical = _decode_map(self._caps)

    def host(self) -> str:
        return "codex"

    def capabilities(self) -> Capabilities:
        return self._caps

    def encode_input(self, inp: Input) -> dict:
        _require_native_or_close(self._caps, "codex", inp.event)

        out: dict = dict(inp.extra or {})
        out["hook_event_name"] = inp.event
        out["session_id"] = inp.session_id
        out["cwd"] = inp.cwd
        if inp.event == "PreToolUse":
            out["tool_name"] = inp.tool_name
            out["tool_input"] = inp.tool_input
        elif inp.event == "PostToolUse":
            out["tool_name"] = inp.tool_name
            out["tool_input"] = inp.tool_input
            out["tool_response"] = inp.tool_response
        elif inp.event == "UserPromptSubmit":
            out["prompt"] = inp.prompt
        return out

    def decode_input(self, raw: dict) -> Input:
        m = _as_dict(raw, "codex input")
        host_event_name = m.get("hook_event_name")
        if not isinstance(host_event_name, str) or host_event_name == "":
            raise SchemaError("codex input: missing hook_event_name")
        event = self._host_to_canonical.get(host_event_name)
        if event is None:
            raise UnsupportedEventError("codex", host_event_name)

        extra = {k: v for k, v in m.items() if k not in _KNOWN_INPUT_KEYS}

        inp = Input(
            event=event,
            session_id=m.get("session_id") if isinstance(m.get("session_id"), str) else "",
            cwd=m.get("cwd") if isinstance(m.get("cwd"), str) else "",
            extra=extra,
        )
        if isinstance(m.get("tool_name"), str):
            inp.tool_name = m["tool_name"]
        if isinstance(m.get("tool_input"), dict):
            inp.tool_input = m["tool_input"]
        if isinstance(m.get("tool_response"), dict):
            inp.tool_response = m["tool_response"]
        if isinstance(m.get("prompt"), str):
            inp.prompt = m["prompt"]
        return inp

    # encode_decision mirrors claude's shape switch: a blank event or
    # PreToolUse use the tool-gate shape; PostToolUse uses the top-level
    # decision/reason shape; SessionStart allows with the empty envelope
    # (or context metadata) and blocks with the same top-level shape.
    # UserPromptSubmit and Stop allow with the empty envelope; codex
    # defines no dedicated block shape for either, so block on those two
    # is unsupported, same as claude.
    def encode_decision(self, d: Decision) -> tuple[object, int]:
        if d.event:
            _require_native_or_close(self._caps, "codex", d.event)
        exit_code = _exit_for(self._caps, d.action)
        event = d.event or "PreToolUse"
        if event == "PreToolUse":
            return _encode_pre_tool_use_decision(d), exit_code
        if event == "PostToolUse":
            return _encode_generic_block_decision(d, "PostToolUse"), exit_code
        if event == "SessionStart":
            return _encode_session_start_decision(d), exit_code
        if event in ("UserPromptSubmit", "Stop"):
            if d.action in ("allow", "warn"):
                return {}, exit_code
            raise UnsupportedActionError(f"codex decision {d.action} for {event}")
        raise UnsupportedEventError("codex", str(d.event))

    def decode_decision(self, stdout: object, exit_code: int) -> Decision:
        if _is_empty_stdout(stdout):
            return Decision(action=_action_for_exit(self._caps, exit_code))
        m = _as_dict(stdout, "codex decision")
        decision, recognized = _decode_known_shape(m)
        if not recognized:
            return Decision(action=_action_for_exit(self._caps, exit_code))
        return decision


def _require_native_or_close(caps: Capabilities, host: str, event: str) -> None:
    lvl, ok = level(caps, event)
    if not ok or lvl not in ("native", "close"):
        raise UnsupportedEventError(host, event)


def _decode_map(caps: Capabilities) -> dict[str, str]:
    m: dict[str, str] = {}
    for p in caps.native:
        if p.host_event in m:
            raise SchemaError(
                f"{caps.host}: host event {p.host_event} maps to both {m[p.host_event]} and {p.event} under native"
            )
        m[p.host_event] = p.event
    seen_close: dict[str, str] = {}
    for p in caps.close or []:
        if p.host_event in seen_close:
            raise SchemaError(
                f"{caps.host}: host event {p.host_event} maps to both "
                f"{seen_close[p.host_event]} and {p.event} under close"
            )
        seen_close[p.host_event] = p.event
        if p.host_event in m:
            continue
        m[p.host_event] = p.event
    return m


def _encode_pre_tool_use_decision(d: Decision) -> dict:
    if d.action == "allow":
        return {}
    if d.action == "rewrite":
        return {
            "hookSpecificOutput": {
                "hookEventName": "PreToolUse",
                "permissionDecision": "allow",
                "updatedInput": d.rewrite,
            }
        }
    if d.action in ("block", "warn"):
        native = _NATIVE_ACTION.get(d.action)
        if native is None:
            raise UnsupportedActionError(f"codex cannot express action {d.action!r}")
        return {
            "hookSpecificOutput": {
                "hookEventName": "PreToolUse",
                "permissionDecision": native,
                "permissionDecisionReason": d.message or "",
            }
        }
    raise UnsupportedActionError(f"codex cannot express action {d.action!r}")


def _encode_generic_block_decision(d: Decision, event: str) -> dict:
    if d.action in ("allow", "warn"):
        return {}
    if d.action == "block":
        return {"decision": "block", "reason": d.message or ""}
    raise UnsupportedActionError(f"codex decision {d.action} for {event}")


def _encode_session_start_decision(d: Decision) -> dict:
    if d.action in ("allow", "warn"):
        with_context = _encode_session_start_context(d.metadata)
        return with_context if with_context is not None else {}
    if d.action == "block":
        return {"decision": "block", "reason": d.message or ""}
    raise UnsupportedActionError(f"codex decision {d.action} for SessionStart")


def _encode_session_start_context(metadata: dict | None) -> dict | None:
    if not metadata:
        return None
    out: dict = {}
    additional_context = metadata.get("additional_context")
    if isinstance(additional_context, str) and additional_context != "":
        out["hookSpecificOutput"] = {"additionalContext": additional_context}
    system_message = metadata.get("system_message")
    if isinstance(system_message, str) and system_message != "":
        out["systemMessage"] = system_message
    if isinstance(metadata.get("continue"), bool):
        out["continue"] = metadata["continue"]
    return out if out else None


# _decode_known_shape mirrors claude's, minus the PermissionRequest
# branch: codex has no PermissionRequest decision channel.
def _decode_known_shape(m: dict) -> tuple[Decision, bool]:
    d = Decision(action="allow")
    recognized = False

    hso = m.get("hookSpecificOutput") if isinstance(m.get("hookSpecificOutput"), dict) else None
    if hso is not None:
        hook_event_name = hso.get("hookEventName")
        if isinstance(hook_event_name, str) and hook_event_name != "":
            d.event = hook_event_name
        if "permissionDecision" in hso:
            recognized = True
            if not d.event:
                d.event = "PreToolUse"
            pd = hso.get("permissionDecision")
            if pd == "deny":
                d.action = "block"
            elif pd == "ask":
                d.action = "warn"
            reason = hso.get("permissionDecisionReason")
            d.message = reason if isinstance(reason, str) else ""
            if "updatedInput" in hso:
                d.action = "rewrite"
                d.rewrite = hso["updatedInput"]
        additional_context = hso.get("additionalContext")
        if isinstance(additional_context, str) and additional_context != "":
            recognized = True
            d.event = "SessionStart"
            d.action = "allow"
            _set_metadata(d, "additional_context", additional_context)

    system_message = m.get("systemMessage")
    if isinstance(system_message, str) and system_message != "":
        recognized = True
        d.event = "SessionStart"
        _set_metadata(d, "system_message", system_message)
    if isinstance(m.get("continue"), bool):
        recognized = True
        d.event = "SessionStart"
        _set_metadata(d, "continue", m["continue"])
    if m.get("decision") == "block":
        recognized = True
        d.action = "block"
        reason = m.get("reason")
        d.message = reason if isinstance(reason, str) else ""
    return d, recognized


def _set_metadata(d: Decision, key: str, value: object) -> None:
    if d.metadata is None:
        d.metadata = {}
    d.metadata[key] = value


def _exit_for(caps: Capabilities, action: str) -> int:
    codes = _host_exit_codes(caps)
    if action == "allow":
        return codes.allow
    if action == "warn":
        return codes.warn if codes.warn is not None else codes.allow
    if action == "block":
        return codes.block
    if action == "rewrite":
        return codes.rewrite if codes.rewrite is not None else codes.allow
    return codes.allow


# _action_for_exit: codex's host.yaml (like claude's) leaves warn unset,
# defaulting it to the same value as allow, so an unset warn code can
# never win a match here unless it is explicitly declared and distinct.
def _action_for_exit(caps: Capabilities, exit_code: int) -> str:
    codes = _host_exit_codes(caps)
    if exit_code == codes.block:
        return "block"
    if codes.rewrite is not None and exit_code == codes.rewrite:
        return "rewrite"
    if codes.warn is not None and codes.warn != codes.allow and exit_code == codes.warn:
        return "warn"
    return "allow"


def _host_exit_codes(caps: Capabilities):
    h = _require_host(caps.host)
    if h.exit_codes is None:
        raise SchemaError(f"{caps.host}: no exit_codes in host.yaml")
    return h.exit_codes


def _require_host(name: str):
    h = get(name)
    if h is None:
        raise SchemaError(f"host {name} not found")
    return h


def _as_dict(v: object, what: str) -> dict:
    if not isinstance(v, dict):
        raise SchemaError(f"{what}: expected a JSON object")
    return v


def _is_empty_stdout(stdout: object) -> bool:
    if stdout is None or stdout == "":
        return True
    if isinstance(stdout, str):
        return stdout.strip() == ""
    return False


def make_codec() -> CodexCodec:
    return CodexCodec()

File: ts/src/hooks/hosts/codex.ts
import { loadCapabilities, level, type Capabilities } from "../../capabilities";
import type { Event } from "../../events";
import { get } from "../../registry";
import { UnsupportedActionError, UnsupportedEventError, SchemaError } from "../../errors";
import type { Codec } from "../codec";
import type { Input } from "../input";
import type { Decision } from "../decision";

/**
 * The Codex CLI codec. Codex's handler contract mirrors Claude Code's:
 * stdin JSON envelope, stdout JSON decision, exit code. Codex's host-side
 * event names also equal the canonical names.
 */
class CodexCodec implements Codec {
  private readonly caps: Capabilities;
  private readonly hostToCanonical: Map<string, Event>;

  constructor() {
    this.caps = loadCapabilities("codex");
    this.hostToCanonical = decodeMap(this.caps);
  }

  host(): string {
    return "codex";
  }

  capabilities(): Capabilities {
    return this.caps;
  }

  encodeInput(input: Input): unknown {
    requireNativeOrClose(this.caps, "codex", input.event);

    const out: Record<string, unknown> = { ...(input.extra ?? {}) };
    out.hook_event_name = input.event;
    out.session_id = input.sessionId;
    out.cwd = input.cwd;
    switch (input.event) {
      case "PreToolUse":
        out.tool_name = input.toolName;
        out.tool_input = input.toolInput;
        break;
      case "PostToolUse":
        out.tool_name = input.toolName;
        out.tool_input = input.toolInput;
        out.tool_response = input.toolResponse;
        break;
      case "UserPromptSubmit":
        out.prompt = input.prompt;
        break;
      default:
        break;
    }
    return out;
  }

  decodeInput(raw: unknown): Input {
    const m = asRecord(raw, "codex input");
    const hostEventName = m.hook_event_name;
    if (typeof hostEventName !== "string" || hostEventName === "") {
      throw new SchemaError("codex input: missing hook_event_name");
    }
    const event = this.hostToCanonical.get(hostEventName);
    if (event === undefined) {
      throw new UnsupportedEventError("codex", hostEventName);
    }

    const known = new Set(["hook_event_name", "session_id", "cwd", "tool_name", "tool_input", "tool_response", "prompt"]);
    const extra: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(m)) {
      if (!known.has(k)) extra[k] = v;
    }

    const input: Input = {
      event,
      sessionId: typeof m.session_id === "string" ? m.session_id : "",
      cwd: typeof m.cwd === "string" ? m.cwd : "",
      extra,
    };
    if (typeof m.tool_name === "string") input.toolName = m.tool_name;
    if (isRecord(m.tool_input)) input.toolInput = m.tool_input;
    if (isRecord(m.tool_response)) input.toolResponse = m.tool_response;
    if (typeof m.prompt === "string") input.prompt = m.prompt;
    return input;
  }

  // encodeDecision mirrors claude's shape switch: a blank event or
  // PreToolUse use the tool-gate shape; PostToolUse uses the top-level
  // decision/reason shape; SessionStart allows with the empty envelope
  // (or context metadata) and blocks with the same top-level shape.
  // UserPromptSubmit and Stop allow with the empty envelope; codex
  // defines no dedicated block shape for either, so block on those two
  // is unsupported, same as claude.
  encodeDecision(d: Decision): { stdout: unknown; exit: number } {
    if (d.event !== undefined && d.event !== "") {
      requireNativeOrClose(this.caps, "codex", d.event);
    }
    const exit = exitFor(this.caps, d.action);
    // `||`, not `??`: Go's `switch d.Event { case "", axon.EventPreToolUse:`
    // treats an explicit empty string exactly like an omitted one. `??`
    // only substitutes on undefined/null and would let an explicit ""
    // fall through to the default/unsupported-event branch.
    switch (d.event || "PreToolUse") {
      case "PreToolUse":
        return { stdout: encodePreToolUseDecision(d), exit };
      case "PostToolUse":
        return { stdout: encodeGenericBlockDecision(d, "PostToolUse"), exit };
      case "SessionStart":
        return { stdout: encodeSessionStartDecision(d), exit };
      case "UserPromptSubmit":
      case "Stop":
        if (d.action === "allow" || d.action === "warn") return { stdout: {}, exit };
        throw new UnsupportedActionError(`codex decision ${String(d.action)} for ${d.event}`);
      default:
        throw new UnsupportedEventError("codex", String(d.event));
    }
  }

  decodeDecision(stdout: unknown, exit: number): Decision {
    if (isEmptyStdout(stdout)) {
      return { action: actionForExit(this.caps, exit) };
    }
    const m = asRecord(stdout, "codex decision");
    const { decision, recognized } = decodeKnownShape(m);
    if (!recognized) {
      return { action: actionForExit(this.caps, exit) };
    }
    return decision;
  }
}

function requireNativeOrClose(caps: Capabilities, host: string, event: Event): void {
  const { level: lvl, ok } = level(caps, event);
  if (!ok || (lvl !== "native" && lvl !== "close")) {
    throw new UnsupportedEventError(host, event);
  }
}

function decodeMap(caps: Capabilities): Map<string, Event> {
  const m = new Map<string, Event>();
  for (const p of caps.native) {
    if (m.has(p.hostEvent)) {
      throw new SchemaError(`${caps.host}: host event ${p.hostEvent} maps to both ${m.get(p.hostEvent)} and ${p.event} under native`);
    }
    m.set(p.hostEvent, p.event);
  }
  const seenClose = new Map<string, Event>();
  for (const p of caps.close ?? []) {
    if (seenClose.has(p.hostEvent)) {
      throw new SchemaError(`${caps.host}: host event ${p.hostEvent} maps to both ${seenClose.get(p.hostEvent)} and ${p.event} under close`);
    }
    seenClose.set(p.hostEvent, p.event);
    if (m.has(p.hostEvent)) continue;
    m.set(p.hostEvent, p.event);
  }
  return m;
}

const NATIVE_ACTION: Partial<Record<Decision["action"], string>> = { allow: "allow", warn: "ask", block: "deny" };

function encodePreToolUseDecision(d: Decision): unknown {
  switch (d.action) {
    case "allow":
      return {};
    case "rewrite":
      return { hookSpecificOutput: { hookEventName: "PreToolUse", permissionDecision: "allow", updatedInput: d.rewrite } };
    case "block":
    case "warn": {
      const native = NATIVE_ACTION[d.action];
      if (!native) throw new UnsupportedActionError(`codex cannot express action ${String(d.action)}`);
      return {
        hookSpecificOutput: {
          hookEventName: "PreToolUse",
          permissionDecision: native,
          permissionDecisionReason: d.message ?? "",
        },
      };
    }
    default:
      throw new UnsupportedActionError(`codex cannot express action ${String(d.action)}`);
  }
}

function encodeGenericBlockDecision(d: Decision, event: string): unknown {
  switch (d.action) {
    case "allow":
    case "warn":
      return {};
    case "block":
      return { decision: "block", reason: d.message ?? "" };
    default:
      throw new UnsupportedActionError(`codex decision ${String(d.action)} for ${event}`);
  }
}

function encodeSessionStartDecision(d: Decision): unknown {
  switch (d.action) {
    case "allow":
    case "warn": {
      const withContext = encodeSessionStartContext(d.metadata);
      return withContext ?? {};
    }
    case "block":
      return { decision: "block", reason: d.message ?? "" };
    default:
      throw new UnsupportedActionError(`codex decision ${String(d.action)} for SessionStart`);
  }
}

function encodeSessionStartContext(metadata: Record<string, unknown> | undefined): Record<string, unknown> | undefined {
  if (!metadata || Object.keys(metadata).length === 0) return undefined;
  const out: Record<string, unknown> = {};
  if (typeof metadata.additional_context === "string" && metadata.additional_context !== "") {
    out.hookSpecificOutput = { additionalContext: metadata.additional_context };
  }
  if (typeof metadata.system_message === "string" && metadata.system_message !== "") {
    out.systemMessage = metadata.system_message;
  }
  if (typeof metadata.continue === "boolean") {
    out.continue = metadata.continue;
  }
  return Object.keys(out).length === 0 ? undefined : out;
}

// decodeKnownShape mirrors claude's, minus the PermissionRequest branch:
// codex has no PermissionRequest decision channel.
function decodeKnownShape(m: Record<string, unknown>): { decision: Decision; recognized: boolean } {
  const d: Decision = { action: "allow" };
  let recognized = false;

  const hso = isRecord(m.hookSpecificOutput) ? m.hookSpecificOutput : undefined;
  if (hso) {
    if (typeof hso.hookEventName === "string" && hso.hookEventName !== "") {
      // hookEventName is untrusted wire data (JSON from the host
      // process), not a value this codec constructs -- same trust
      // boundary as Event's other wire-loaded assignments (events.ts,
      // capabilities.ts). Type annotation only; the runtime value and
      // control flow are unchanged.
      d.event = hso.hookEventName as Event;
    }
    if ("permissionDecision" in hso) {
      recognized = true;
      if (d.event === undefined) d.event = "PreToolUse";
      const pd = hso.permissionDecision;
      if (pd === "deny") d.action = "block";
      else if (pd === "ask") d.action = "warn";
      d.message = typeof hso.permissionDecisionReason === "string" ? hso.permissionDecisionReason : "";
      if ("updatedInput" in hso) {
        d.action = "rewrite";
        d.rewrite = hso.updatedInput;
      }
    }
    if (typeof hso.additionalContext === "string" && hso.additionalContext !== "") {
      recognized = true;
      d.event = "SessionStart";
      d.action = "allow";
      setMetadata(d, "additional_context", hso.additionalContext);
    }
  }
  if (typeof m.systemMessage === "string" && m.systemMessage !== "") {
    recognized = true;
    d.event = "SessionStart";
    setMetadata(d, "system_message", m.systemMessage);
  }
  if (typeof m.continue === "boolean") {
    recognized = true;
    d.event = "SessionStart";
    setMetadata(d, "continue", m.continue);
  }
  if (m.decision === "block") {
    recognized = true;
    d.action = "block";
    d.message = typeof m.reason === "string" ? m.reason : "";
  }
  return { decision: d, recognized };
}

function setMetadata(d: Decision, key: string, value: unknown): void {
  if (!d.metadata) d.metadata = {};
  d.metadata[key] = value;
}

function exitFor(caps: Capabilities, action: Decision["action"]): number {
  const codes = hostExitCodes(caps);
  switch (action) {
    case "allow":
      return codes.allow;
    case "warn":
      return codes.warn ?? codes.allow;
    case "block":
      return codes.block;
    case "rewrite":
      return codes.rewrite ?? codes.allow;
    default:
      return codes.allow;
  }
}

// actionForExit: codex's host.yaml (like claude's) leaves warn unset,
// defaulting it to the same value as allow, so an unset warn code can
// never win a match here unless it is explicitly declared and distinct.
function actionForExit(caps: Capabilities, exit: number): Decision["action"] {
  const codes = hostExitCodes(caps);
  if (exit === codes.block) return "block";
  if (codes.rewrite !== undefined && exit === codes.rewrite) return "rewrite";
  if (codes.warn !== undefined && codes.warn !== codes.allow && exit === codes.warn) return "warn";
  return "allow";
}

function hostExitCodes(caps: Capabilities): { allow: number; block: number; warn?: number; rewrite?: number } {
  const h = requireHost(caps.host);
  if (!h.exitCodes) throw new SchemaError(`${caps.host}: no exit_codes in host.yaml`);
  return h.exitCodes;
}

function requireHost(name: string) {
  const h = get(name);
  if (!h) throw new SchemaError(`host ${name} not found`);
  return h;
}

function isRecord(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

function asRecord(v: unknown, what: string): Record<string, unknown> {
  if (!isRecord(v)) throw new SchemaError(`${what}: expected a JSON object`);
  return v;
}

function isEmptyStdout(stdout: unknown): boolean {
  if (stdout === undefined || stdout === null || stdout === "") return true;
  if (typeof stdout === "string") return stdout.trim() === "";
  return false;
}

export function makeCodec(): Codec {
  return new CodexCodec();
}
